python_modules/dagster/dagster/_core/execution/retries.py

```diff
@@ -69,3 +80,66 @@ def mark_attempt(self, key: str) -> None:
 
     def snapshot_attempts(self) -> Mapping[str, int]:
         return dict(self._attempts)
+
+
+def auto_reexecution_should_retry_run(instance: "DagsterInstance", run: DagsterRun):
+    """Determines if a run will be retried by the automatic reexcution system.
+    A run will retry if:
+    - it is failed.
+    - the number of max allowed retries is > 0 (max retries can be set via system setting or run tag).
+    - there have not already been >= max_retries retries for the run.
+
+    If the run failure reason was a step failure and the retry_on_asset_or_op_failure tag/system setting is set to false,
+    a warning message will be logged and the run will not be retried.
+
+    We determine how many retries have been launched for the run by looking at the size of the run group
+    (the set of runs that have the same root_run_id and the run with root_run_id). Since manually launched retries are
+    part of the run group, this means that if a user launches a manual retry of run A and then this function
+    is called because a retry for run A launched by the auto-reexecution system failed, the manual retry will be
+    counted toward max_retries.
+
+    There is potential that one "extra" retry will be launched by the automatic reexecution system
+    since manual retries could be happening in parallel with automatic retries. Here is
+    an illustrative example:
```

Comment:
this is pretty unlikely to happen in practice right? seems like things have to happen at just the right time for a relatively rare sequence of events - if so the comment here could maybe make that clearer